scvi/module/base/_base_module.py

```diff
@@ -343,3 +358,155 @@ def create_predictive(
+        Tuple[jnp.ndarray, jnp.ndarray, LossRecorder],
+    ]:
+        """
+        Forward pass through the network.
+
+        Parameters
+        ----------
+        tensors
+            tensors to pass through
+        get_inference_input_kwargs
+            Keyword args for `_get_inference_input()`
+        get_generative_input_kwargs
+            Keyword args for `_get_generative_input()`
+        inference_kwargs
+            Keyword args for `inference()`
+        generative_kwargs
+            Keyword args for `generative()`
+        loss_kwargs
+            Keyword args for `loss()`
+        compute_loss
+            Whether to compute loss on forward pass. This adds
+            another return value.
+        """
+        return _generic_forward(
+            self,
+            tensors,
+            inference_kwargs,
+            generative_kwargs,
+            loss_kwargs,
+            get_inference_input_kwargs,
+            get_generative_input_kwargs,
+            compute_loss,
+        )
+
+    @abstractmethod
+    def _get_inference_input(self, tensors: Dict[str, jnp.ndarray], **kwargs):
+        """Parse tensors dictionary for inference related values."""
+
+    @abstractmethod
+    def _get_generative_input(
+        self,
+        tensors: Dict[str, jnp.ndarray],
+        inference_outputs: Dict[str, jnp.ndarray],
+        **kwargs,
+    ):
+        """Parse tensors dictionary for generative related values."""
+
+    @abstractmethod
+    def inference(
+        self,
+        *args,
+        **kwargs,
+    ) -> Dict[str, Union[jnp.ndarray, Distribution]]:
+        """
+        Run the inference (recognition) model.
+
+        In the case of variational inference, this function will perform steps related to
+        computing variational distribution parameters. In a VAE, this will involve running
+        data through encoder networks.
+
+        This function should return a dictionary with str keys and :class:`~jnp.ndarray` values.
+        """
+        pass
+
+    @abstractmethod
+    def generative(
+        self, *args, **kwargs
+    ) -> Dict[str, Union[jnp.ndarray, Distribution]]:
+        """
+        Run the generative model.
+
+        This function should return the parameters associated with the likelihood of the data.
+        This is typically written as :math:`p(x|z)`.
+
+        This function should return a dictionary with str keys and :class:`~jnp.ndarray` values.
+        """
+        pass
+
+    @abstractmethod
+    def loss(self, *args, **kwargs) -> LossRecorder:
+        """
+        Compute the loss for a minibatch of data.
+
+        This function uses the outputs of the inference and generative functions to compute
+        a loss. This many optionally include other penalty terms, which should be computed here.
+
+        This function should return an object of type :class:`~scvi.module.base.LossRecorder`.
+        """
+        pass
+
+
+def _generic_forward(
```

Comment:
where else is this used?

Reply:
? Only inside this file, both by jax call and pytorch forward

Reply:
Oh ok missed the pytorch one. Just wanted to warn against refactoring too early.